ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionConverter.kt

```diff
@@ -0,0 +1,130 @@
+        }
+        for ((taskId, value) in docResponses) {
+            try {
+                putResponse(taskId, job, value, responses)
+            } catch (e: DataStoreException) {
+                Timber.e(e, "Task $taskId in remote db in submission $submissionId")
+            }
+        }
+        return responses.build()
+    }
+
+    private fun putResponse(
+        taskId: String, job: Job, obj: Any, responses: ResponseMap.Builder
+    ) {
+        val task = job.getTask(taskId).orElseThrow { DataStoreException("Not defined in task") }
+        when (task.type) {
+            Task.Type.PHOTO, Task.Type.TEXT_FIELD -> putTextResponse(taskId, obj, responses)
+            Task.Type.MULTIPLE_CHOICE -> putMultipleChoiceResponse(
+                taskId,
+                task.multipleChoice,
+                obj,
+                responses
+            )
+            Task.Type.NUMBER -> putNumberResponse(taskId, obj, responses)
+            Task.Type.DATE -> putDateResponse(taskId, obj, responses)
+            Task.Type.TIME -> putTimeResponse(taskId, obj, responses)
+            else -> throw DataStoreException("Unknown type " + task.type)
+        }
+    }
+
+    private fun putNumberResponse(taskId: String, obj: Any, responses: ResponseMap.Builder) {
+        val value = DataStoreException.checkType(Double::class.java, obj) as Double
+        NumberResponse.fromNumber(value.toString())
+            .ifPresent { r: Response -> responses.putResponse(taskId, r) }
+    }
+
+    private fun putTextResponse(taskId: String, obj: Any, responses: ResponseMap.Builder) {
+        val value = DataStoreException.checkType(String::class.java, obj) as String
+        TextResponse.fromString(value.trim { it <= ' ' })
+            .ifPresent { r: Response -> responses.putResponse(taskId, r) }
+    }
+
+    private fun putDateResponse(taskId: String, obj: Any, responses: ResponseMap.Builder) {
+        val value = DataStoreException.checkType(Timestamp::class.java, obj) as Timestamp
+        DateResponse.fromDate(value.toDate())
+            .ifPresent { r: Response -> responses.putResponse(taskId, r) }
+    }
+
+    private fun putTimeResponse(taskId: String, obj: Any, responses: ResponseMap.Builder) {
+        val value = DataStoreException.checkType(Timestamp::class.java, obj) as Timestamp
+        TimeResponse.fromDate(value.toDate())
+            .ifPresent { r: Response -> responses.putResponse(taskId, r) }
+    }
+
+    private fun putMultipleChoiceResponse(
+        taskId: String, multipleChoice: MultipleChoice?, obj: Any, responses: ResponseMap.Builder
+    ) {
+        val values = DataStoreException.checkType(MutableList::class.java, obj) as List<*>
+        values.forEach { v: Any? ->
+            DataStoreException.checkType(String::class.java, v as Any)
+        }
+        MultipleChoiceResponse.fromList(multipleChoice, values as List<String>)
+            .ifPresent { r: Response -> responses.putResponse(taskId, r) }
```

Comment:
"r: Response" could probably be omitted as well, unless you think it's needed for readability